### ISSUE TITLE
assignment error not detected as expected

### ISSUE THREAD

**Issue by sileht** [python/mypy#14800]
**Bug Report**

assignment error not detected as expected.

**To Reproduce**

```python
import typing


def mypy_broken(_type: typing.Literal["1", "2", "3"]) -> None:
    if _type == "1":
        count = 0
    elif _type == "2":
        count = 5
    elif _type == "3":
        count = None
    else:
        # No error returned while we expected:
        # Incompatible types in assignment
        # (expression has type "str", variable has type "int|None")
        count = "a_string"

    # show: int | None
    reveal_type(count)


def mypy_broken_v2(_type: typing.Literal["1", "2", "3"]) -> None:
    count: int | None
    if _type == "1":
        count = 0
    elif _type == "2":
        count = 5
    elif _type == "3":
        count = None
    else:
        # No error returned while we expected:
        # Incompatible types in assignment
        # (expression has type "str", variable has type "int|None")
        count = "a_string"

    # show: int | None
    reveal_type(count)


def mypy_working(_type: typing.Literal["1", "2", "3"]) -> None:
    if _type == "1":
        count = 0
    elif _type == "2":
        count = 5
    else:
        # error is here correctly reported : Incompatible types in assignment
        # (expression has type "str", variable has type "int")
        count = "a_string"

    # show: int
    reveal_type(count)
```

**Expected Behavior**

**Actual Behavior**

**Your Environment**

- Mypy version used: mypy 1.0.1 (compiled: yes)
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files):
```
[tool.mypy]
strict = true
exclude = [".venv"]
files = ["xxxx"]
show_error_codes = true
plugins = ["pydantic.mypy", "sqlalchemy.ext.mypy.plugin"]

[[tool.mypy.overrides]]
module = [
       "deepdiff.*",
       "vcr.*",
       "voluptuous.*",
       "msgpack.*",
       "bs4.*",
       "graphql_utils",
       "markdownify",
       "asgi_lifespan",
       "authlib.*",
       "imia.*",
       "starsessions.*",
       "sqlalchemy_utils.*",
]
ignore_missing_imports = true
```
- Python version used: 3.11.2

**Comment by JelleZijlstra** [comment]
Mypy doesn't check code it detects as unreachable. You should use the `--warn-unreachable` option to get warned about this.

**Comment by sileht** [comment]
Thank! I'm a bit surprise it's not part of strict mode but ok.

**Comment by hauntsaninja** [comment]
I think it's not part of `--strict` because it technically doesn't affect soundness.

I find it's also not uncommon to run into cases where `--warn-unreachable` is a little noisy (but really depends on your codebase)

**Comment by hauntsaninja** [comment]
Definitely the most useful option that's not in `--strict`!
